src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/CheckForPoison.cs

```diff
@@ -213,6 +227,20 @@ internal IEnumerable<PoisonedFileEntry> GetPoisonedFiles(IEnumerable<string> ini
             return poisons;
         }
 
+        private IEnumerable<string> GetAllNonShippingPackages()
+        {
+            if (NonShippingPackagesListFiles != null)
+            {
+                return NonShippingPackagesListFiles
+                    .SelectMany(item => File.ReadAllLines(item.ItemSpec))
+                    .Distinct();
```

Comment:
I think you want to include a ToList() invocation here otherwise the entire linq query is re-evaluated on every iteration of the while loop in `GetPoisonedFiles`

Reply:
Fixed in commit 4.